File: chefbar/motion.py
"""Signaal motion helpers for ChefBar (GTK3).

DESIGN.md: 100ms press, 180ms hover/select, 280ms panel/dialog.
Only opacity fades here — no bounce, pulse loops, or scroll theater.
Honors GtkSettings gtk-enable-animations (reduced-motion proxy).
"""

from __future__ import annotations

import weakref
from typing import TYPE_CHECKING, Any, Callable

# Signaal motion scale (ms)
PRESS_MS = 100
HOVER_MS = 180
PANEL_MS = 280

_FADE_STEPS = 8

# Per-window animation generation. Starting any fade bumps the token so
# timeouts from an earlier fade become stale and stop touching the window
# (e.g. a fade-out tick must not hide a window reopened by fade_in).
_generations: weakref.WeakKeyDictionary[Any, int] = weakref.WeakKeyDictionary()

if TYPE_CHECKING:
    from gi.repository import Gtk


def _gtk() -> tuple[Any, Any]:
    import gi

    gi.require_version("Gtk", "3.0")
    from gi.repository import GLib, Gtk

    return GLib, Gtk


def motion_enabled() -> bool:
    """Return False when the desktop asks for reduced / no animations."""
    _GLib, Gtk = _gtk()
    settings = Gtk.Settings.get_default()
    if settings is None:
        return True
    try:
        return bool(settings.get_property("gtk-enable-animations"))
    except Exception:  # noqa: BLE001
        return True


def _begin(window: Any) -> int:
    """Invalidate any in-flight fade on ``window`` and return a fresh token."""
    try:
        generation = _generations.get(window, 0) + 1
        _generations[window] = generation
    except TypeError:  # window not weak-referenceable / hashable
        return 0
    return generation


def _is_current(window: Any, generation: int) -> bool:
    try:
        return _generations.get(window, 0) == generation
    except TypeError:
        return True


def fade_in(window: Gtk.Window, *, duration_ms: int = PANEL_MS) -> None:
    """Show window with a short opacity ramp (or instantly if motion off).

    Caller may already have called ``show_all`` for layout; we still ramp
    opacity from 0 → 1 when animations are enabled.
    """
    GLib, _Gtk = _gtk()
    generation = _begin(window)
    if not motion_enabled() or duration_ms <= 0:
        window.set_opacity(1.0)
        window.show()
        return

    window.set_opacity(0.0)
    window.show()
    step_ms = max(1, duration_ms // _FADE_STEPS)
    state = {"i": 0}

    def tick() -> bool:
        if not _is_current(window, generation):
            return False
        state["i"] += 1
        frac = min(1.0, state["i"] / _FADE_STEPS)
        window.set_opacity(frac)
        if frac >= 1.0:
            window.set_opacity(1.0)
            return False
        return True

    GLib.timeout_add(step_ms, tick)


def fade_out(
    window: Gtk.Window,
    *,
    duration_ms: int = PANEL_MS,
    on_hidden: Callable[[], None] | None = None,
) -> None:
    """Hide window after a short opacity ramp (or instantly if motion off)."""
    GLib, _Gtk = _gtk()
    generation = _begin(window)

    def finish() -> None:
        window.hide()
        window.set_opacity(1.0)
        if on_hidden is not None:
            on_hidden()

    if not motion_enabled() or duration_ms <= 0:
        finish()
        return

    step_ms = max(1, duration_ms // _FADE_STEPS)
    state = {"i": 0}

    def tick() -> bool:
        if not _is_current(window, generation):
            # A newer fade (usually a reopen) owns the window now.
            return False
        state["i"] += 1
        frac = max(0.0, 1.0 - state["i"] / _FADE_STEPS)
        window.set_opacity(frac)
        if frac <= 0.0:
            finish()
            return False
        return True

    GLib.timeout_add(step_ms, tick)

File: src/motion.rs
//! Signaal motion helpers voor ChefBar (GTK3).
//!
//! Signaal v2 motion-spec: 140ms hover/press, 280ms paneel, 420ms expand.
//! Alleen opacity-fades: geen bounce, pulse-loops of scroll-theater.
//! Respecteert gtk-enable-animations (reduced-motion proxy).

use gtk::glib::ControlFlow;
use gtk::prelude::*;
use std::collections::HashMap;
use std::sync::Mutex;

/// Devin v2 duur-ladder (`tokens.css`): `--dur-fast` / `--dur-med` /
/// `--dur-slow`. Elke timing hieronder moet één van deze drie zijn — geen
/// tussenwaarden, anders loopt de chrome uit de pas met de designfile.
pub const DUR_FAST_MS: u32 = 140;
pub const DUR_MED_MS: u32 = 280;
pub const DUR_SLOW_MS: u32 = 420;

pub const PRESS_MS: u32 = DUR_FAST_MS;
pub const HOVER_MS: u32 = DUR_FAST_MS;
pub const PANEL_MS: u32 = DUR_MED_MS;
/// Drawer is een pane-slide en volgt dus dezelfde duur als het paneel
/// (was 160ms — naast de ladder).
pub const DRAWER_MS: u32 = DUR_MED_MS;
/// Palette-overlay is een reveal, niet een pane (was 100ms — naast de ladder).
pub const OVERLAY_MS: u32 = DUR_FAST_MS;
const FADE_STEPS: u32 = 8;

/// Per-window animation generation: elke start-fade bumped de token zodat
/// timeouts van een eerdere fade de window niet meer aanraken (een fade-out
/// tick mag een heropend venster niet verbergen).
static GENERATIONS: std::sync::LazyLock<Mutex<HashMap<usize, u32>>> =
    std::sync::LazyLock::new(|| Mutex::new(HashMap::new()));

fn window_key(window: &gtk::Window) -> usize {
    window.as_ptr() as usize
}

fn begin(window: &gtk::Window) -> u32 {
    let mut map = GENERATIONS.lock().unwrap();
    let next = map.get(&window_key(window)).copied().unwrap_or(0) + 1;
    map.insert(window_key(window), next);
    next
}

fn is_current(window: &gtk::Window, generation: u32) -> bool {
    GENERATIONS
        .lock()
        .unwrap()
        .get(&window_key(window))
        .copied()
        == Some(generation)
}

pub fn motion_enabled() -> bool {
    gtk::Settings::default()
        .map(|settings| settings.property::<bool>("gtk-enable-animations"))
        .unwrap_or(true)
}

/// Toon window met een korte opacity-ramp (of instant als motion uit staat).
pub fn fade_in(window: &gtk::Window, duration_ms: u32) {
    let generation = begin(window);
    if !motion_enabled() || duration_ms == 0 {
        window.set_opacity(1.0);
        window.show();
        return;
    }
    window.set_opacity(0.0);
    window.show();
    let step_ms = std::cmp::max(1, duration_ms / FADE_STEPS);
    let window = window.clone();
    let mut step = 0u32;
    gtk::glib::timeout_add_local(
        std::time::Duration::from_millis(step_ms as u64),
        move || {
            if !is_current(&window, generation) {
                return ControlFlow::Break;
            }
            step += 1;
            let frac = (step as f64 / FADE_STEPS as f64).min(1.0);
            window.set_opacity(frac);
            if frac >= 1.0 {
                window.set_opacity(1.0);
                ControlFlow::Break
            } else {
                ControlFlow::Continue
            }
        },
    );
}

/// Verberg window na een korte opacity-ramp (of instant).
pub fn fade_out(window: &gtk::Window, duration_ms: u32) {
    let generation = begin(window);
    let window_clone = window.clone();

    fn finish(window: &gtk::Window) {
        window.hide();
        window.set_opacity(1.0);
    }

    if !motion_enabled() || duration_ms == 0 {
        finish(&window_clone);
        return;
    }
    let step_ms = std::cmp::max(1, duration_ms / FADE_STEPS);
    let mut step = 0u32;
    gtk::glib::timeout_add_local(
        std::time::Duration::from_millis(step_ms as u64),
        move || {
            if !is_current(&window_clone, generation) {
                return ControlFlow::Break;
            }
            step += 1;
            let frac = (1.0 - step as f64 / FADE_STEPS as f64).max(0.0);
            window_clone.set_opacity(frac);
            if frac <= 0.0 {
                finish(&window_clone);
                ControlFlow::Break
            } else {
                ControlFlow::Continue
            }
        },
    );
}

/// Slide drawer in/uit (`DRAWER_MS`). Wrapper om fade_in/fade_out.
/// TODO: echte translate-animatie als GTK reveal-infrastructuur beschikbaar is;
/// voor nu fade — compileerbaar en visueel consistent met panel.
pub fn slide_drawer(window: &gtk::Window, open: bool) {
    if open {
        fade_in(window, DRAWER_MS);
    } else {
        fade_out(window, DRAWER_MS);
    }
}

/// Fade palette-overlay in/uit (`OVERLAY_MS`). Wrapper om fade_in/fade_out.
pub fn fade_overlay(window: &gtk::Window, show: bool) {
    if show {
        fade_in(window, OVERLAY_MS);
    } else {
        fade_out(window, OVERLAY_MS);
    }
}

/// Bereken de zichtbare rij-breedte aan de hand van een fractie (0..1) en de
/// track-breedte in pixels — voor de usage-bars.
pub fn fill_width_px(frac: f64, track_width: i32) -> i32 {
    let frac = frac.clamp(0.0, 1.0);
    (track_width.max(0) as f64 * frac).round() as i32
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn timings_blijven_op_de_devin_ladder() {
        let ladder = [DUR_FAST_MS, DUR_MED_MS, DUR_SLOW_MS];
        for (naam, ms) in [
            ("PRESS_MS", PRESS_MS),
            ("HOVER_MS", HOVER_MS),
            ("PANEL_MS", PANEL_MS),
            ("DRAWER_MS", DRAWER_MS),
            ("OVERLAY_MS", OVERLAY_MS),
        ] {
            assert!(
                ladder.contains(&ms),
                "{naam} = {ms}ms staat niet op de ladder 140/280/420"
            );
        }
    }

    #[test]
    fn ladder_matcht_tokens_css() {
        assert_eq!((DUR_FAST_MS, DUR_MED_MS, DUR_SLOW_MS), (140, 280, 420));
    }

    #[test]
    fn fill_width_clamps() {
        assert_eq!(fill_width_px(0.5, 100), 50);
        assert_eq!(fill_width_px(2.0, 100), 100);
        assert_eq!(fill_width_px(-1.0, 100), 0);
    }
}
